File: 자료구조/자료구조.cpp
#include <deque>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>

using namespace std;

int main() {
  // stack
  stack<int> s;
  s.push(1);
  s.push(2);
  s.push(3);
  s.top(); // 3
  s.pop();
  s.top(); // 2

  // queue
  queue<int> q;
  q.push(1);
  q.push(2);
  q.push(3);
  q.front(); // 1
  q.pop();
  q.front(); // 2

  // deque
  deque<int> dq;
  dq.push_back(10);
  dq.push_front(20);
  dq.push_back(30);
  dq.front(); // 20
  dq.back();  // 30
  dq.pop_front();
  dq.front(); // 10

  // priority_queue - 최대 힙
  priority_queue<int> pq;
  pq.push(50);
  pq.push(20);
  pq.push(40);
  pq.top(); // 50

  // priority_queue - 최소 힙
  priority_queue<int, vector<int>, greater<int>> min_pq;
  min_pq.push(50);
  min_pq.push(20);
  min_pq.push(40);
  min_pq.top(); // 20

  // hash map
  unordered_map<string, int> umap;
  umap["banana"] = 20;
  umap["orange"] = 30;
  umap["apple"] = 10;

  for (auto it : umap) {
    cout << it.first << " " << it.second << endl;
  }
  // banana 20
  // orange 30
  // apple 10

  // map
  map<string, int> map;
  map["banana"] = 20;
  map["apple"] = 10;
  map["orange"] = 30;

  for (auto it : map) {
    cout << it.first << " " << it.second << endl;
  }

  // apple 10
  // banana 20
  // orange 30

  // set
  set<int> set;
  set.insert(100);
  set.insert(50);
  set.insert(150);
  set.insert(50); // 중복 추가

  for (auto it : set) {
    cout << it << " ";
  }
  // 50 100 150

  return 0;
}

File: 자료구조/자료구조.js
class Stack {
    constructor(){
        this.arr = [];
    }
    push(a){
        this.arr.push(a);
    }
    pop(){
        this.arr.pop();
    }
    top(){
        return this.arr[this.arr.length-1];
    }
    empty(){
        return this.arr.length === 0;
    }
}

class Queue {
    constructor(){
        this.arr = [];
    }
    push(a){
        this.arr.push(a);
    }
    pop(){
        this.arr.shift();
    }
    top(){
        return this.arr[0];
    }
    empty(){
        return this.arr.length === 0;
    }
}

class Dequeue {
    constructor(){
        this.arr = [];
    }
    push_front(a){
        this.arr.unshift(a);
    }
    push_back(a){
        this.arr.push(a);
    }
    pop_front(){
        this.arr.pop();
    }
    pop_back(){
        this.arr.shift();
    }
    front(){
        return this.arr[0];
    }
    back(){
        return this.arr[this.arr.length-1];
    }
    empty(){
        return this.arr.length === 0;
    }
}

class PriorityQueue {
    constructor() {
        this.arr = [];
    }
    push(a) {
        this.arr.push(a);
        this.heapifyUp();
    }
    pop() {
        if (this.arr.length === 0) return null;
        if (this.arr.length === 1) return this.arr.pop();
        
        let top = this.arr[0];
        this.arr[0] = this.arr.pop();
        this.heapifyDown();
        return top;
    }
    top() {
        return this.arr[0];
    }
    size() {
        return this.arr.length;
    }
    heapifyUp() {
        let index = this.arr.length - 1;
         
        while (index > 0) {
            let parent = Math.floor((index - 1) / 2);
            if (this.arr[parent] <= this.arr[index]) break;
            
            [this.arr[index], this.arr[parent]] =
                [this.arr[parent], this.arr[index]];
            
            index = parent;
        }
    }
    heapifyDown() {
        let index = 0;
        let length = this.arr.length;
        
        while(1) {
            let left_child = index * 2 + 1;
            let right_child = index * 2 + 2;
            let smallest = index;
            
            if(left_child < length &&
               this.arr[left_child] < this.arr[smallest])
                smallest = left_child;
            if(right_child < length &&
               this.arr[right_child] < this.arr[smallest])
                smallest = right_child;
            
            if(index == smallest) break;
            
            [this.arr[index], this.arr[smallest]] = 
                [this.arr[smallest], this.arr[index]];
            
            index = smallest;
        }
    }
}


// map

let map = new Map();

map.set("apple", 100);
map.set("banana", 200);
console.log(map.get("apple")); // 100
console.log(map.has("banana")); // true

map.delete("banana");
console.log(map.has("banana")); // false

console.log([...map.keys()]); // ["apple"]
console.log([...map.values()]); // [100]

// set

let mySet = new Set();

mySet.add(10);
mySet.add(20);
mySet.add(10); // 중복 저장 X

console.log(mySet.has(10)); // true
console.log(mySet.has(30)); // false

mySet.delete(10);
console.log(mySet.has(10)); // false

console.log([...mySet]); // [20]
